lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -522,49 +522,65 @@ class MongoClientInterface {
+     * master only if the returned version is greater than the stored
+     * one. Caveat: this function is not optimized for multiple
+     * updates to the same objName, a batch would be more suited to
+     * avoid the parallel attempts to update the master.
      */
     putObjectVerCase4(c, bucketName, objName, objVal, params, log, cb) {
         const vObjName = formatVersionKey(objName, params.versionId);
-        const mst = generatePHDVersion(params.versionId);
-        return c.bulkWrite([{
-            updateOne: {
-                filter: {
-                    _id: vObjName,
-                },
-                update: {
-                    $set: {
-                        _id: vObjName,
-                        value: objVal,
-                    },
-                },
-                upsert: true,
-            },
+        c.update({
+            _id: vObjName,
         }, {
-            updateOne: {
-                filter: {
-                    _id: objName,
-                },
-                update: {
-                    _id: objName, value: mst,
-                },
-                upsert: true,
-            },
-        }], {
-            ordered: 1,
+            _id: vObjName,
+            value: objVal,
+        }, {
+            upsert: true,
         }, err => {
             if (err) {
-                log.error('putObjectVerCase4: error putting object version', {
-                    error: err.message,
-                });
+                log.error(
+                    'putObjectVerCase4: error upserting object version',
+                    { error: err.message });
                 return cb(errors.InternalError);
             }
-            return cb(null, `{"versionId": "${objVal.versionId}"}`);
+            this.getLatestVersion(c, objName, log, (err, mstObjVal) => {
+                if (err) {
+                    log.error('getLatestVersion: getting latest version',
+                              { error: err.message });
+                    return cb(err);
+                }
+                // eslint-disable-next-line
+                c.update({
+                    _id: objName,
+                    'value.versionId': {
+                        $gt: mstObjVal.versionId,
+                    },
+                }, {
+                    _id: objName,
+                    value: mstObjVal,
+                }, {
+                    upsert: true,
+                }, err => {
+                    if (err) {
+                        // we accept that the update fails if
+                        // condition is not met, meaning that a more
+                        // recent master was already in place
+                        if (err.code !== 11000) {
+                            log.error(
+                                'putObjectVerCase4: error upserting master',
+                                { error: err.message });
+                            return cb(errors.InternalError);
```

Comment:
Do we want to return `InternalError` in this case because technically it's a known situation not an actual error?

Reply:
This update does not return an error if the object is replaced (even if the content is not changed).
The master update may return an error if the condition is not met, hence the check for error 11000.

Reply:
I mean the InternalError will catch other errors than an actual overwrite (e.g. I/O errors).

Reply:
Ok.